File: csharp/StrangeCharacters/Characters/Character.cs
namespace Characters;

public class Character
{
    public string FirstName { get; }
    public string? LastName { get; }
    public bool IsMonster { get; }

    public Character? Nemesis { get; private set; }

    public HashSet<Character> children = new();
    public HashSet<Character> siblings = new();
    public HashSet<Character> parents = new();

    public Character(string firstName, string? lastName=null, bool isMonster=false)
    {
        FirstName = firstName;
        LastName = lastName;
        IsMonster = isMonster;
    }

    public void AddChild(Character child)
    {
        children.Add(child);
        child.parents.Add(this);
        foreach (var character in children)
        {
            if (child.FirstName != character.FirstName)
            {
                child.siblings.Add(character);
                character.siblings.Add(child);
            }
        }
    }

    public void SetNemesis(Character monster)
    {
        this.Nemesis = monster;
    }

    public override string ToString()
    {
        return $"Character({FirstName}, {LastName}, isMonster:{IsMonster})";
    }

    protected bool Equals(Character other)
    {
        return FirstName == other.FirstName;
    }

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((Character)obj);
    }

    public override int GetHashCode()
    {
        return FirstName.GetHashCode();
    }
}

File: scala/src/main/scala/characters/Character.scala
package characters

import scala.collection.mutable

class Character(val firstName: String, val lastName: Option[String] = None, val isMonster: Boolean = false):

  private var _nemesis: Option[Character] = None

  val children: mutable.Set[Character] = new mutable.HashSet[Character]()
  val siblings: mutable.Set[Character] = new mutable.HashSet[Character]()
  val parents: mutable.Set[Character] = new mutable.HashSet[Character]()

  def addChild(child: Character): Unit =
    children.add(child)
    child.parents.add(this)
    children.foreach { character =>
      if child.firstName != character.firstName then
        child.siblings.add(character)
        character.siblings.add(child)
    }

  def nemesis: Option[Character] = _nemesis
  def nemesis_=(newNemesis: Character): Unit = _nemesis = Some(newNemesis)

  override def toString: String = s"Character($firstName, $lastName, isMonster:$isMonster)"

  protected def equals(other: Character): Boolean = firstName == other.firstName

  override def equals(obj: Any): Boolean =
    if null == obj then false
    else if obj.getClass != this.getClass then false
    else equals(obj.asInstanceOf[Character])

  override def hashCode(): Int = firstName.hashCode
